Automattic/wp-calypso
# Checkout and Payment Methods: Remove useTotal from payment methods

In https://github.com/Automattic/wp-calypso/pull/81807 we removed a bunch of exports from the `@automattic/composite-checkout` package. Many were unused, some were leaky abstractions, and others didn't fit the vision of the package in its current form. One of the exports that was not removed was `useTotal` because it is still used in many places.

When `CheckoutProvider` was first implemented, the idea was that it would be provided a list of line items and a total in a standardized format and would then display those line items in a particular way automatically. However, as time passed it became clear that consumers needed full control over how line items were rendered, making the whole standardized line item feature a liability and a second source of truth. In https://github.com/Automattic/wp-calypso/pull/63720 we finally removed `useLineItems` from all payment methods. However, `useTotal` remained because most payment methods used it to display the total inside their submit button.

Fortunately, the shopping cart knows the total and that can be provided to each payment method directly.

## Proposed Changes

This PR replaces `useTotal` in every payment method that still uses it with the shopping cart total instead. It also de-duplicates the code of each payment method to require a React Node for its submit button content (effectively using dependency injection for the button content). 

This was a necessary refactor because some payment methods might be used outside the context of a shopping cart (eg: the "add payment method" page) and have their own submit button text. However, the "Pay X" text which (now) requires the shopping cart was the default, which means that `useShoppingCart` had to be called for every payment method. This would throw an error if the payment method was not inside a `ShoppingCartProvider` even if it didn't plan on using the cart at all. (It has the side benefit of also not needing us to move all payment methods into the `client` directory to have access to `useCartKey` which is required by `useShoppingCart`.)

Once this is complete we'll be able to remove `useTotal` entirely in https://github.com/Automattic/wp-calypso/pull/82750

## Screenshots (before and after)

Jetpack checkout, ready state.

<img width="619" alt="Screenshot 2023-10-10 at 11 41 50 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/790ef9bc-26f9-40b1-997e-d5a17ad5ce46">

WPCom checkout, ready state.

<img width="598" alt="Screenshot 2023-10-10 at 11 42 14 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/5eda0bf9-21fb-4de1-9371-c0ecc7d6a9e7">

WPCom checkout, submitting state.

<img width="596" alt="Screenshot 2023-10-10 at 11 46 57 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/38ef852e-6ced-4e96-a5ce-1d40adf0b919">

WPCom checkout, loading state.

<img width="514" alt="Screenshot 2023-10-11 at 12 04 30 AM" src="https://github.com/Automattic/wp-calypso/assets/2036909/be7f6634-c2da-4f41-a5be-79db2583ccd1">

Change payment method form, ready state.

<img width="697" alt="Screenshot 2023-10-10 at 11 42 51 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/bb059b5c-3c2a-4520-98a6-b657892de8eb">

## Testing Instructions

This PR should not change any behavior or UI so the goal of testing is to make sure that nothing has changed from production.

Visit checkout with a product in the cart and try selecting different payment methods. Verify that their submit buttons look the same before and after this PR.

Use an account with at least one active subscription. Visit `/me/purchases`, click on the subscription, and click "Change payment method" (or "Add payment method" if none is assigned). Try selecting different payment methods and verify that the submit buttons look the same before and after this PR.
